File: src/scenes/Pause.cpp
#include "Pause.hpp"
#include "CppInvaders.hpp"

#define PAUSE_SYMBOL_BLINK 0.5
#define TIME_TO_RESUME 3

PauseScene::PauseScene(Scene *gameplay$) {
    gameplay = gameplay$;
    resuming = false;
    timer = 0;
    pause_symbol = true;
}

PauseScene::~PauseScene() {}

void PauseScene::process_event(const Pico_Event &event) {
    switch (event.type) {
    case PICO_EVENT_KEY_DOWN:
        switch (event.key.keysym.sym) {
            case PICOK_ESCAPE: // start resume countdown or cancel it
                resuming = !resuming;
                timer = 0;
                pause_symbol = true;
                break;
        }
        break;
    }
}

void PauseScene::update(float delta) {
    timer += delta;

    if (resuming && timer >= TIME_TO_RESUME) {
        CppInvaders::get().scene = gameplay;
        delete this;
        return;
    }

    if (!resuming && timer >= PAUSE_SYMBOL_BLINK) {
        timer = 0;
        pause_symbol = !pause_symbol;
        return;
    }
}

void PauseScene::draw() const {
    if (gameplay) {
        gameplay->draw();
    }

    pico_push();
    // pause menu dim effect
    pico_set_color_draw((Pico_Color){0, 0, 0});
    pico_set_alpha(204);
    Pico_Rel_Rect dim_r = { '%', {0.0f, 0.0f, 1.0f, 1.0f}, PICO_ANCHOR_NW, NULL };
    pico_output_draw_rect(&dim_r);
    pico_pop();

    pico_set_color_draw(PICO_COLOR_WHITE);
    Pico_Rel_Rect txt_r = { '#', {14.0f, 4.0f, 0.0f, 1.0f}, PICO_ANCHOR_N, NULL };

    if (resuming) {
        char fmt[16];
        snprintf(fmt, sizeof(fmt), "%1d", 3 - (int)timer);
        pico_output_draw_text(fmt, &txt_r);
    } else if (pause_symbol) {
        pico_output_draw_text("||", &txt_r);
    }
}

File: lua/scenes/Pause.lua
local class = require "class"

local PauseScene = class.new()

function PauseScene:init(game_state, gameplay)
    self.game_state = game_state
    self.gameplay = gameplay
    self.resuming = false
    self.timer = 0
    self.pause_symbol = true
end

function PauseScene:process_event(event)
    if event.tag == 'key.dn' and event.key == 'Escape' then
        self.resuming = not self.resuming
        self.timer = 0
        self.pause_symbol = true
    end
end

function PauseScene:update(delta)
    self.timer = self.timer + delta

    if self.resuming and self.timer >= 3 then
        self.game_state.set_scene(self.gameplay)
        return
    end

    if not self.resuming and self.timer >= 0.5 then
        self.timer = 0
        self.pause_symbol = not self.pause_symbol
        return
    end
end

function PauseScene:draw()
    if self.gameplay then
        self.gameplay:draw()
    end

    pico.push()
    pico.set.color.draw({r=0, g=0, b=0})
    pico.set.alpha(204)
    local dim_r = { '%', x=0.0, y=0.0, w=1.0, h=1.0, anchor='NW' }
    pico.output.draw.rect(dim_r)
    pico.pop()

    pico.set.color.draw("white")
    local txt_r = { '#', x=14.0, y=4.0, w=0.0, h=1.0, anchor='N' }

    if self.resuming then
        local fmt = tostring(3 - math.floor(self.timer))
        pico.output.draw.text(fmt, txt_r)
    elseif self.pause_symbol then
        pico.output.draw.text("||", txt_r)
    end
end

return PauseScene
